### PR TITLE
Implement CBOR parsing

### PR DESCRIPTION
**Pull request by felix-gohla**
This PR implements parsing of CBOR-encoded data and can parse the `authenticatorGetInfo` response.

It uses the [cb0r](https://github.com/quartzjer/cb0r) library.

### CODE REVIEW

File: src/info.c [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+}
+
+/**
+ * @brief Parse the versions array from the CBOR response.
+ * 
+ * @param element The element in the array.
+ * @param ci User-passed argument (here: CBOR info).
+ * @return int FIDO_OK if versions could be parsed.
+ */
+int cbor_info_decode_versions(const cb0r_t element, void *ci) {
```
Comment (sirkrypt0) [inline src/info.c:109]:
```suggestion
static int cbor_info_decode_versions(const cb0r_t element, void *ci) {
```

File: src/info.c [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+}
+
+/**
+ * @brief Parse the extensions array from the CBOR response.
+ * 
+ * @param element The element in the array.
+ * @param ci User-passed argument (here: CBOR info).
+ * @return int FIDO_OK if extensions could be parsed.
+ */
+int cbor_info_decode_extensions(const cb0r_t element, void *ci) {
```
Comment (sirkrypt0) [inline src/info.c:138]:
```suggestion
static int cbor_info_decode_extensions(const cb0r_t element, void *ci) {
```

File: src/info.c [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+/**
+ * @brief Parse the options map from the CBOR response.
+ * 
+ * @param key The key in the map.
+ * @param value The value in the map.
+ * @param ci User-passed argument (here: CBOR info).
+ * @return int FIDO_OK if extensions could be parsed.
+ */
+int cbor_info_decode_options(const cb0r_t key, const cb0r_t value, void *ci) {
```
Comment (sirkrypt0) [inline src/info.c:168]:
```suggestion
static int cbor_info_decode_options(const cb0r_t key, const cb0r_t value, void *ci) {
```

File: src/info.c [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+            break;
+        default:
+            // Just ignore.
+            break;
+    }
+    return FIDO_OK;
+}
+
+
+static int cbor_info_decode_transport(const cb0r_t element, void *arg) {
```
Comment (sirkrypt0) [inline src/info.c:256]: All the others have comments.

File: src/io.c [hunk 5]
```diff
@@ -32,6 +32,7 @@ int fido_rx(fido_dev_t *d, const uint8_t cmd, void *buf, const size_t len) {
         return (FIDO_ERR_INVALID_ARGUMENT);
     }
 
+    // Errors are less than 0.
```
Comment (frcroth) [inline src/io.c:35]:
```suggestion
    // Values below 0 are errors.
```

File: include/info.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    fido_cbor_pin_protocol_t protocols;      /* supported pin protocols */
+    // TODO: How to support authenticator preference?
+    fido_cbor_algorithm_t    algorithms;     /* list of supported algorithms */
+    uint64_t                 maxcredcntlst;  /* max credentials in list */
+    uint64_t                 maxcredidlen;   /* max credential ID length */
+    uint64_t                 fwversion;      /* firmware version */
+    uint64_t                 maxcredbloblen; /* max credBlob length */
+    uint64_t                 maxlargeblob;   /* max largeBlob array length */
+} fido_cbor_info_t;
+
```
Comment (frcroth) [inline include/info.h:112]: Doc here?

File: src/info.c [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+        return FIDO_ERR_CBOR_UNEXPECTED_TYPE;
+    }
+
+    return cbor_iter_map(element, cbor_info_decode_algorithm_entry, arg);
+}
+
+
+static int parse_info_reply_entry(const cb0r_t key, const cb0r_t value, void *arg) {
+    if (key->type != CB0R_INT || key->value > UINT8_MAX) {
+        // Just ignore the entry according to...
```
Comment (sirkrypt0) [inline src/info.c:322]: According to?

File: src/info.c [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    // This should always be a map.
+    if (!cb0r_read(msg+1, msglen-1, &map) || map.type != CB0R_MAP) {
+        return  FIDO_ERR_CBOR_UNEXPECTED_TYPE;
+    }
+
+    // The next step parses the response.
+    return cbor_iter_map(&map, &parse_info_reply_entry, ci);
+}
+
+int fido_dev_get_cbor_info_wait(fido_dev_t *dev, fido_cbor_info_t *ci) {
```
Comment (frcroth) [inline src/info.c:401]: Doc would be nice.

File: src/nfc.c [hunk 9]
```diff
@@ -13,19 +13,11 @@
...
-#define AVR_PROGMEM
-#endif
-
 #define TX_CHUNK_SIZE 240
 
-static const uint8_t aid[]                              = { 0xa0, 0x00, 0x00, 0x06, 0x47, 0x2f, 0x00, 0x01 };
-static const uint8_t fido_version_u2f[] AVR_PROGMEM     = { 'U', '2', 'F', '_', 'V', '2' };
-static const uint8_t fido_version_fido2[] AVR_PROGMEM   = { 'F', 'I', 'D', 'O', '_', '2', '_', '0' };
+static const uint8_t aid[]                                  = { 0xa0, 0x00, 0x00, 0x06, 0x47, 0x2f, 0x00, 0x01 };
+static const uint8_t fido_version_u2f[] PROGMEM_MARKER      = { 'U', '2', 'F', '_', 'V', '2' };
```
Comment (sirkrypt0) [inline src/nfc.c:19]: Could use strings `""` here as well.

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

File: src/info.c (hunk 8 above)
Comment (felix-gohla) [reply src/info.c:401]: Should be in the header.

File: external/cb0r/src/cb0r.c [hunk 10]
```diff
@@ -0,0 +1,425 @@
+// by jeremie miller - 2015-2017
+// public domain UNLICENSE, contributions/improvements welcome via github at https://github.com/quartzjer/cb0r
```
[the remark this answers is not in the document]
Comment (felix-gohla) [reply external/cb0r/src/cb0r.c:2]: We added our copyright notice below, would this suffice?

File: src/dev.c [hunk 11]
```diff
@@ -12,6 +12,75 @@
...
+            }
+        }
+    }
+    */
+}
+
+static void fido_dev_set_protocol_flags(fido_dev_t *dev, const fido_cbor_info_t *info) {
+    // TODO
+    /*
+    const uint8_t  *ptr = fido_cbor_info_protocols_ptr(info);
```
Comment (felix-gohla) [inline src/dev.c:58]: Those could be implemented in this PR.

**Review by sirkrypt0 — APPROVED** [review]

File: src/dev.c (hunk 11 above)
Comment (frcroth) [reply src/dev.c]: Done

Comment (sirkrypt0) [reply src/dev.c]: @felix-gohla Please check if you are fine with the new implementation.

File: src/info.c (hunk 7 above)
Comment (felix-gohla) [reply src/info.c]: See <https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#message-encoding>

> If map keys are present that an implementation does not understand, they MUST be ignored. Note that this enables additional fields to be used as new features are added without breaking existing implementations.

**Comment by frcroth** [comment]
@felix-gohla Will you merge?

**Pull request merged**